### ISSUE TITLE
AV: Explore the use of Window.postMessage() to replace autoplay on media-player

### ISSUE THREAD

**Issue by HarveyPeachey** [bbc/simorgh#4670]
**Is your feature request related to a problem? Please describe.**
From the investigation on https://github.com/bbc/psammead/issues/2614 we found that removing autoplay and invoking `.play()` directly on the media player is the best method that could prevent users requiring two clicks to play media on certain browsers. However, we ran into issues with cross-origin and with a forthcoming change of the domain that our media embeds are going to be served on, see https://github.com/bbc/simorgh/issues/3970 and JIRA ticket `TA-1586` for more information, this is going to persist however there is a possible solution.

Additional benefits would include:

- Ability to access elements within the iframe for testing purposes.
- Ability to control SMP from Simorgh.

**Describe the solution you'd like**
From the investigation, we want to explore the use of [`Window.postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) to dispatch a message from our `psammead-media-player` component to the av-embed iframe. The av-embed iframe would subscribe to this event and then action by playing the media player.

We need to make sure the code we write to apply this is thoroughly checked because if it isn't this could cause security issues, see [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage#Security_concerns).

**Describe alternatives you've considered**
The only other way to combat the two-click bug is to apply the muted attribute to the `iFrame` for the browsers which don't allow autoplay, see https://github.com/bbc/psammead/issues/2614

**Testing notes**

- Unit Tests

- [x] This feature is expected to need manual testing.

**Additional context**
Discussion about changes to the AV embed domain https://github.com/bbc/simorgh/issues/3970

**Comment by HarveyPeachey** [comment]
Here is a repository for testing the use of `postMessage` between iFrames: https://github.com/HarveyPeachey/postmessage-av-concept

**Comment by HarveyPeachey** [comment]
**Here are my findings:**

`Window.postMessage()` is a powerful method for safely communicating between the iframes we use for our AV media player. Unfortunately, the method explored for this issue doesn't seem to work.

In this concept [repository](https://github.com/HarveyPeachey/postmessage-av-concept) I've attempted to simulate the current behaviour of how our AV player works in terms of embedded iframes. 

I managed to use `Window.postMessage()` to safely communicate cross-origin by having the embedded video running on one node.js server, and the simorgh article mock page on another node.js server by using different port numbers for localhost. I then use `postMessage()` to trigger `play()` on the video player. However, chrome wouldn't allow this as it violates it's autoplay policy because the user hasn't interacted with the document iframe window:

![media-play-failed](https://user-images.githubusercontent.com/4706195/69145119-d31d4500-0ac4-11ea-984b-4358aaa61700.gif)

To try and combat this I thought we could pass the `mouseClick` event from when the button was pressed through `postMessage()` and use this to satisfy user interaction on the domain however this is not possible as you can't pass objects that contain functions, see [The structured clone algorithm
](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm).

Another solution that I thought might've been possible is triggering the playback of the video with a button that lives on the same domain as the video embed on the domain `localhost:8080/testdomain`. That way chrome could see that the domain has been interacted with by the user, thus theoretically allowing the use of `play()`. But this is not the case as it seems the detection for user interaction is encapsulated in the iframe itself:
![play-iframe-trick](https://user-images.githubusercontent.com/4706195/69145780-5a1eed00-0ac6-11ea-89a0-57ff02fde985.gif)

And here's an example of when the play button is triggered and a user clicks on the iframe document before `play()` is invoked on the video, which allows playback as it satisfies chromes autoplay policy:
![play-dom-interaction](https://user-images.githubusercontent.com/4706195/69146478-dcf47780-0ac7-11ea-9815-656c58807f55.gif)


**Summary:**
From my investigation into this `Window.postMessage()` is not going to allow us to invoke `play()` and get around autoplay policy issues. Although `Window.postMessage()` is very powerful and can be used for other functionality that we may need for the AV media-player. If anyone has any questions or other solution ideas let me know below.

**Comment by RayNjeri** [comment]
This is quite a detailed investigation and thanks for adding all the necessary links to add more context to the issue, Maybe one question i would have is have you explored the option of adding an invisible `iframe` and setting `allow=”autoplay”` before the AV element?

**Comment by HarveyPeachey** [comment]
> This is quite a detailed investigation and thanks for adding all the necessary links to add more context to the issue, Maybe one question i would have is have you explored the option of adding an invisible `iframe` and setting `allow=”autoplay”` before the AV element?

We have `autoplay` already for our AV player. The main issue was that chrome on mobile wouldn't allow the use of `autoplay`. So I investigated invoking `play()` directly on the video as a workaround.

**Comment by ryanmccombe** [comment]
Nice investigation 👍

Any workaround we find to "trick" browsers into autoplaying video is naturally going to be a hack.  Browsers have already found and patched dozens of work arounds - even if we do manage to find another, that will get fixed eventually too, and our double-click bug will be back

The best solution to this is likely going to be embracing the double click, and coming up with a UX approach that makes it friendly.  eg, instead of the first click being a play button that doesn't play, we could make it a "load video" button.

**Comment by 12** [comment]
In extension to what @ryanmccombe said, more browsers are blocking websites from autoplaying video altogether - which is a shame considering the actual UX is not autoplay. 

Thanks for doing this investigation @HarveyPeachey

**Comment by HarveyPeachey** [comment]
I've also learnt that Chrome allows `autoplay` on sites that have been added to a user's home screen. So if we ever offer a pop-up option to add the site to a user's home screen in the future, this could mitigate the double click issue for some users.

**Comment by HarveyPeachey** [comment]
> Nice investigation 👍
> 
> Any workaround we find to "trick" browsers into autoplaying video is naturally going to be a hack. Browsers have already found and patched dozens of work arounds - even if we do manage to find another, that will get fixed eventually too, and our double-click bug will be back
> 
> The best solution to this is likely going to be embracing the double click, and coming up with a UX approach that makes it friendly. eg, instead of the first click being a play button that doesn't play, we could make it a "load video" button.

I agree with this approach. 

But if we are to make the UX be a "load media button" or something similar, we're going to have to decide whether we make that UX only appear for browsers which don't work with autoplay, or have that UX for all browser versions and remove autoplay altogether.

**Comment by lizcameron** [comment]
UX team is reviewing the placeholder/messaging at the moment to see if that improves the UX for two clicks. 
We still need to consider whether this would/could be introduced on all browsers or just on those without autoplay, or whether the messaging works for either.

**Comment by HarveyPeachey** [comment]
@lizcameron Am I okay to close this investigation now? And discussion about how we proceed can happen outside of this issue?

**Comment by jamesdonoh** [comment]
This investigation is really nice work @HarveyPeachey - thanks for the links, screenshots and including the code you used.

The most pragmatic thing to do IMO is to update the UX as described above in a way that would satisfy all users regardless of whether the AV will automatically play for them or not. Knowing in advance which group they fall into is likely to be difficult - while we could theoretically have certain rules these will quickly go out of date as browser vendors' approaches evolve.

Closing this as we've finished the investigation described above and can create separate issues for any follow-on work.
